packages/flame_tiled/lib/src/renderable_tile_map.dart

```diff
@@ -81,98 +108,148 @@ class RenderableTiledMap {
-      _renderableTileLayers(map).map((e) => _loadImages(map)),
+    Vector2 destTileSize, {
+    Camera? camera,
+  }) async {
+    final renderableLayers = await Future.wait(
+      map.layers.where((layer) => layer.visible).toList().map((layer) {
+        switch (layer.runtimeType) {
+          case TileLayer:
+            return _RenderableTileLayer.load(
+              layer as TileLayer,
+              map,
+              destTileSize,
+            );
+          case ImageLayer:
+            return _RenderableImageLayer.load(layer as ImageLayer, camera);
+
+          default:
+            return _UnrenderableLayer.load(layer);
+        }
+      }),
     );
 
     return RenderableTiledMap(
       map,
-      batchesByLayer,
-      destTileSize,
+      renderableLayers,
+      camera: camera,
     );
   }
 
-  static Iterable<TileLayer> _renderableTileLayers(TiledMap map) {
-    return map.layers.where((layer) => layer.visible).whereType<TileLayer>();
+  /// Handle game resize and propagate it to renderable layers
+  void handleResize(Vector2 canvasSize) {
+    renderableLayers.forEach((rl) {
+      rl.handleResize(canvasSize);
+    });
   }
 
-  static Future<Map<String, SpriteBatch>> _loadImages(TiledMap map) async {
-    final result = <String, SpriteBatch>{};
-
-    await Future.forEach(map.tiledImages(), (TiledImage img) async {
-      final src = img.source;
-      if (src != null) {
-        result[src] = await SpriteBatch.load(src);
-      }
+  /// Rebuilds the cache for rendering
+  void _refreshCache() {
+    renderableLayers.forEach((rl) {
+      rl.refreshCache();
     });
-
-    return result;
   }
 
-  /// Rebuilds the cache for rendering
-  void refreshCache() {
-    batchesByLayer.forEach(
-      (batchMap) => batchMap.values.forEach((batch) => batch.clear()),
-    );
+  /// Renders each renderable layer in the same order specified by the Tiled map
+  void render(Canvas c) {
+    if (_backgroundPaint != null) {
+      c.drawPaint(_backgroundPaint!);
+    }
 
-    _renderableTileLayers(map)
-        .where((e) => e.tileData != null)
-        .forEachIndexed((mapIndex, layer) {
-      return _renderLayer(
-        mapIndex,
-        layer.tileData!,
-        Vector2(layer.offsetX, layer.offsetY),
-      );
+    // paint each layer in reverse order, because the last layers should be
+    // rendered beneath the first layers
+    renderableLayers.where((l) => l.visible).forEach((renderableLayer) {
+      renderableLayer.render(c, camera);
     });
   }
 
-  void _renderLayer(
-    int mapIndex,
-    List<List<Gid>> tileData,
-    Vector2 layerOffset,
+  /// Returns a layer of type [T] with given [name] from all the layers
+  /// of this map. If no such layer is found, null is returned.
+  T? getLayer<T extends Layer>(String name) {
+    return map.layers
+        .firstWhereOrNull((layer) => layer is T && layer.name == name) as T?;
+  }
+}
+
+abstract class _RenderableLayer<T extends Layer> {
```

Comment:
In the future (but that is for a much larger refactor) this should probably be a component, because then it can just have `PositionType.viewport` (or just added on the viewport with the new camera system) and it wouldn't have to have the camera sent in to it.

Reply:
Agreed, I could do that refactor in a follow-up PR, but then there would be no way to use `RenderableTileMap` without using the Component system. Not sure if y'all want to support that